File: external_apps/mobile_ios/mobile_ios/Core/Realtime/WebSocketManager.swift
//
//  WebSocketManager.swift
//  mobile_ios
//
//  Created for Realtime Feature
//

import Foundation
import SocketIO

enum ConnectionState {
    case disconnected
    case connecting
    case connected
    case error(String)
}

class WebSocketManager: ObservableObject {
    static let shared = WebSocketManager()
    
    @Published var connectionState: ConnectionState = .disconnected
    
    private var manager: SocketManager?
    private var socket: SocketIOClient?
    
    private init() {}
    
    func connect(url: String, token: String? = nil) {
        connectionState = .connecting
        
        var config: SocketIOClientConfiguration = [
            .log(false),
            .compress,
            .reconnects(true),
            .reconnectAttempts(5),
            .reconnectWait(1)
        ]
        
        if let token = token {
            config.insert(.connectParams(["token": token]))
        }
        
        guard let socketURL = URL(string: url) else {
            connectionState = .error("Invalid URL")
            return
        }
        
        manager = SocketManager(socketURL: socketURL, config: config)
        socket = manager?.defaultSocket
        
        socket?.on(clientEvent: .connect) { [weak self] _, _ in
            self?.connectionState = .connected
        }
        
        socket?.on(clientEvent: .disconnect) { [weak self] _, _ in
            self?.connectionState = .disconnected
        }
        
        socket?.on(clientEvent: .error) { [weak self] data, _ in
            let error = data.first as? String ?? "Connection error"
            self?.connectionState = .error(error)
        }
        
        socket?.connect()
    }
    
    func disconnect() {
        socket?.disconnect()
        connectionState = .disconnected
    }
    
    func on(event: String, callback: @escaping ([Any]) -> Void) {
        socket?.on(event, callback: callback)
    }
    
    func off(event: String) {
        socket?.off(event)
    }
    
    func emit(event: String, items: [Any]) {
        socket?.emit(event, items)
    }
    
    func isConnected() -> Bool {
        return socket?.status == .connected
    }
    
    // MARK: - Common Events
    func onMessage(callback: @escaping ([String: Any]) -> Void) {
        on(event: "message") { data in
            if let dict = data.first as? [String: Any] {
                callback(dict)
            }
        }
    }
    
    func onNotification(callback: @escaping ([String: Any]) -> Void) {
        on(event: "notification") { data in
            if let dict = data.first as? [String: Any] {
                callback(dict)
            }
        }
    }
    
    func onUserStatusChange(callback: @escaping ([String: Any]) -> Void) {
        on(event: "user_status") { data in
            if let dict = data.first as? [String: Any] {
                callback(dict)
            }
        }
    }
    
    func sendMessage(message: String, recipientId: String) {
        let data: [String: Any] = [
            "message": message,
            "recipientId": recipientId,
            "timestamp": Date().timeIntervalSince1970
        ]
        emit(event: "send_message", items: [data])
    }
}

File: external_apps/mobile_android/app/src/main/java/com/example/mobile_android/data/realtime/WebSocketManager.kt
package com.example.mobile_android.data.realtime

import io.socket.client.IO
import io.socket.client.Socket
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import org.json.JSONObject
import javax.inject.Inject
import javax.inject.Singleton

sealed class ConnectionState {
    object Disconnected : ConnectionState()
    object Connecting : ConnectionState()
    object Connected : ConnectionState()
    data class Error(val message: String) : ConnectionState()
}

@Singleton
class WebSocketManager @Inject constructor() {
    
    private var socket: Socket? = null
    
    private val _connectionState = MutableStateFlow<ConnectionState>(ConnectionState.Disconnected)
    val connectionState: StateFlow<ConnectionState> = _connectionState
    
    private val eventListeners = mutableMapOf<String, MutableList<(JSONObject) -> Unit>>()
    
    fun connect(url: String, token: String? = null) {
        try {
            _connectionState.value = ConnectionState.Connecting
            
            val options = IO.Options().apply {
                reconnection = true
                reconnectionAttempts = 5
                reconnectionDelay = 1000
                token?.let {
                    auth = mapOf("token" to it)
                }
            }
            
            socket = IO.socket(url, options)
            
            socket?.apply {
                on(Socket.EVENT_CONNECT) {
                    _connectionState.value = ConnectionState.Connected
                }
                
                on(Socket.EVENT_DISCONNECT) {
                    _connectionState.value = ConnectionState.Disconnected
                }
                
                on(Socket.EVENT_CONNECT_ERROR) { args ->
                    val error = args.firstOrNull()?.toString() ?: "Connection error"
                    _connectionState.value = ConnectionState.Error(error)
                }
                
                connect()
            }
        } catch (e: Exception) {
            _connectionState.value = ConnectionState.Error(e.message ?: "Unknown error")
        }
    }
    
    fun disconnect() {
        socket?.disconnect()
        socket = null
        _connectionState.value = ConnectionState.Disconnected
    }
    
    fun on(event: String, callback: (JSONObject) -> Unit) {
        socket?.on(event) { args ->
            val data = args.firstOrNull() as? JSONObject ?: JSONObject()
            callback(data)
        }
        
        // Store listener for cleanup
        eventListeners.getOrPut(event) { mutableListOf() }.add(callback)
    }
    
    fun off(event: String) {
        socket?.off(event)
        eventListeners.remove(event)
    }
    
    fun emit(event: String, data: JSONObject) {
        socket?.emit(event, data)
    }
    
    fun isConnected(): Boolean {
        return socket?.connected() == true
    }
    
    // Common event handlers
    fun onMessage(callback: (JSONObject) -> Unit) {
        on("message", callback)
    }
    
    fun onNotification(callback: (JSONObject) -> Unit) {
        on("notification", callback)
    }
    
    fun onUserStatusChange(callback: (JSONObject) -> Unit) {
        on("user_status", callback)
    }
    
    fun sendMessage(message: String, recipientId: String) {
        val data = JSONObject().apply {
            put("message", message)
            put("recipientId", recipientId)
            put("timestamp", System.currentTimeMillis())
        }
        emit("send_message", data)
    }
}
